File: Sep/Day1/maximum-average-pass-ratio.java
// TLE  ::  Time: O(K * N)  ::   Space: O(N)
class Solution {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        int n = classes.length;
        double[] passRatio = new double[n];
        for (int i = 0; i < n; i++) {
            passRatio[i] = (double) classes[i][0] / classes[i][1];
        }

        while (extraStudents-- > 0) {
            double[] updatedRatios = new double[n];
            for (int i = 0; i < n; i++) {
                updatedRatios[i] = (double) (classes[i][0] + 1) / (classes[i][1] + 1);
            }

            int chosenIdx = 0;
            double maxGain = 0.0;
            for (int i = 0; i < n; i++) {
                double gain = updatedRatios[i] - passRatio[i];
                if (gain > maxGain) {
                    maxGain = gain;
                    chosenIdx = i;
                }
            }

            passRatio[chosenIdx] = updatedRatios[chosenIdx];
            classes[chosenIdx][0]++;
            classes[chosenIdx][1]++;
        }

        double total = 0;
        for (double r : passRatio) total += r;
        return total / n;
    }
}





// Optimized  ::  Time: O((K + N) * Log(N))  ::   Space: O(N)
import java.util.*;

class Solution {
    static class ClassData {
        int pass, total;
        ClassData(int p, int t) { pass = p; total = t; }
        double gain() {
            return (double)(pass + 1) / (total + 1) - (double)pass / total;
        }
    }

    public double maxAverageRatio(int[][] classes, int extraStudents) {
        PriorityQueue<ClassData> pq = new PriorityQueue<>((a, b) -> 
            Double.compare(b.gain(), a.gain())
        );

        for (int[] c : classes) pq.add(new ClassData(c[0], c[1]));

        while (extraStudents-- > 0) {
            ClassData top = pq.poll();
            top.pass++;
            top.total++;
            pq.add(top);
        }

        double total = 0;
        for (ClassData c : pq) {
            total += (double)c.pass / c.total;
        }
        return total / classes.length;
    }
}

File: Sep/Day1/maximum-average-pass-ratio.js
// TLE  ::  Time: O(K * N)  ::   Space: O(N)
class Solution {
    maxAverageRatio(classes, extraStudents) {
        let passRatio = classes.map(c => c[0] / c[1]);

        while (extraStudents-- > 0) {
            let updatedRatios = classes.map(c => (c[0] + 1) / (c[1] + 1));

            let chosenIdx = 0;
            let maxGain = 0;
            for (let i = 0; i < updatedRatios.length; i++) {
                let gain = updatedRatios[i] - passRatio[i];
                if (gain > maxGain) {
                    maxGain = gain;
                    chosenIdx = i;
                }
            }

            passRatio[chosenIdx] = updatedRatios[chosenIdx];
            classes[chosenIdx][0]++;
            classes[chosenIdx][1]++;
        }

        let total = passRatio.reduce((a, b) => a + b, 0);
        return total / classes.length;
    }
}






// Optimized  ::  Time: O((K + N) * Log(N))  ::   Space: O(N)
class MaxHeap {
    constructor(compare) {
        this.data = [];
        this.compare = compare;
    }
    push(val) {
        this.data.push(val);
        this._siftUp(this.data.length - 1);
    }
    pop() {
        if (this.size() === 1) return this.data.pop();
        const top = this.data[0];
        this.data[0] = this.data.pop();
        this._siftDown(0);
        return top;
    }
    size() { return this.data.length; }
    _siftUp(i) {
        while (i > 0) {
            let p = Math.floor((i - 1) / 2);
            if (this.compare(this.data[i], this.data[p])) {
                [this.data[i], this.data[p]] = [this.data[p], this.data[i]];
                i = p;
            } else break;
        }
    }
    _siftDown(i) {
        let n = this.data.length;
        while (true) {
            let l = 2 * i + 1, r = 2 * i + 2, largest = i;
            if (l < n && this.compare(this.data[l], this.data[largest])) largest = l;
            if (r < n && this.compare(this.data[r], this.data[largest])) largest = r;
            if (largest !== i) {
                [this.data[i], this.data[largest]] = [this.data[largest], this.data[i]];
                i = largest;
            } else break;
        }
    }
}

class Solution {
    maxAverageRatio(classes, extraStudents) {
        const gain = (p, t) => (p + 1) / (t + 1) - p / t;

        const heap = new MaxHeap((a, b) => a.gain > b.gain);
        for (let [p, t] of classes) {
            heap.push({ pass: p, total: t, gain: gain(p, t) });
        }

        while (extraStudents-- > 0) {
            let top = heap.pop();
            top.pass++;
            top.total++;
            top.gain = gain(top.pass, top.total);
            heap.push(top);
        }

        let total = 0;
        while (heap.size()) {
            let { pass, total: t } = heap.pop();
            total += pass / t;
        }
        return total / classes.length;
    }
}
